nixos-rebuild: init as package

zimbatm opened the pull request:
###### Motivation for this change

Move the nixos-rebuild script from the nixos distribution into the
package set.

This allows non-nixos users to run the script as well. For example,
deploying a remote machine with:

    nixos-rebuild --target-host root@hostname --build-host root@hostname

If that idea is interesting, potentially move all the other nixos
commands to packages.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

zimbatm committed:
nixos-*: init as package

Move all the nixos-* scripts from the nixos distribution as real
packages in the pkgs/ package set.

This allows non-nixos users to run the script as well. For example,
deploying a remote machine with:

    nixos-rebuild --target-host root@hostname --build-host root@hostname

GrahamcOfBorg commented:
<!--REQUEST_ID=61864d5e-d07d-4805-8ed9-fea92dd4d495-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=e1046808-672b-4e88-b00a-2949dcef8c30)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/9kipmsxjyjgj5ddlxzqqrs6l9wlmii57-busybox-1.29.3' from 'https://cache.nixos.org'...
copying path '/nix/store/hik4v1b8a8h3mga9g0ghx481kysjmqp5-nix-2.1.3-man' from 'https://cache.nixos.org'...
copying path '/nix/store/ag088hzpkdlzd9lgxvi094bxicvcn32k-libatomic_ops-7.6.6' from 'https://cache.nixos.org'...
copying path '/nix/store/5fvlfkwzfl3a89apvvyaz7652clrcf76-aws-sdk-cpp-1.5.17' from 'https://cache.nixos.org'...
copying path '/nix/store/bwlaq8bvbpyqsfw9a5rgz7wr09rdpikm-curl-7.59.0' from 'https://cache.nixos.org'...
copying path '/nix/store/4ri26b7qgh53s65iqyj50sj4pvknl1qv-libsodium-1.0.16' from 'https://cache.nixos.org'...
copying path '/nix/store/846brfjv2rw3gk5yq744316qpkmk84jl-boehm-gc-7.6.8' from 'https://cache.nixos.org'...
copying path '/nix/store/crg1xmgsyqfcxk0wp5pycrzn32sbwwjh-nix-2.1.3' from 'https://cache.nixos.org'...
building '/nix/store/73qcwchc0aa4f9y35ywdrvy2l13lhhpx-nixos-rebuild.drv'...
/nix/store/q0asgrckwqgk4zl8y2z5v9dl2ws87isy-nixos-rebuild
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=61864d5e-d07d-4805-8ed9-fea92dd4d495-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=e7a1e7ae-2fa4-48e6-a02f-fe7f90797cd6)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
these derivations will be built:
  /nix/store/39vrdzsyw5wax59x6rmdk7hk1n0zlxcz-nixos-rebuild.drv
building '/nix/store/39vrdzsyw5wax59x6rmdk7hk1n0zlxcz-nixos-rebuild.drv'...
/nix/store/qfnl0zskb5sj1br41ndvav1pi0xs718p-nixos-rebuild
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=61864d5e-d07d-4805-8ed9-fea92dd4d495-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=9bf5e0b4-45c6-4855-83a4-e3578f1126b5)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
  /nix/store/6rf5riyf2yd6db8rcbpl3iq34753na0p-aws-sdk-cpp-1.5.17
  /nix/store/7ncbv7a0z43bh0bhp4bl3r0qy79aq8k9-libsodium-1.0.16
  /nix/store/plwp97zfj51hgr3rv80jkqmwa8mdxv79-nixos-rebuild
  /nix/store/s02dcjkyf7f6584k03cv1v6pman9m1vl-curl-7.59.0
copying path '/nix/store/6rf5riyf2yd6db8rcbpl3iq34753na0p-aws-sdk-cpp-1.5.17' from 'https://cache.nixos.org'...
copying path '/nix/store/s02dcjkyf7f6584k03cv1v6pman9m1vl-curl-7.59.0' from 'https://cache.nixos.org'...
copying path '/nix/store/7ncbv7a0z43bh0bhp4bl3r0qy79aq8k9-libsodium-1.0.16' from 'https://cache.nixos.org'...
copying path '/nix/store/1m6749jbggi50mn6jh0k9893377mhyqi-nix-2.1.3' from 'https://cache.nixos.org'...
copying path '/nix/store/plwp97zfj51hgr3rv80jkqmwa8mdxv79-nixos-rebuild' from 'https://cache.nixos.org'...
/nix/store/plwp97zfj51hgr3rv80jkqmwa8mdxv79-nixos-rebuild
```
</p></details>

zimbatm committed:
move the codeName to /.codeName

Make the codeName globally accessible in the repo. The release is not
only for NixOS anymore.

GrahamcOfBorg commented:
<!--REQUEST_ID=02b5a084-6302-4c4d-9268-663e27ee6ca4-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=8f765623-fe3d-4771-89c6-ea508126de50)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
these derivations will be built:
  /nix/store/39vrdzsyw5wax59x6rmdk7hk1n0zlxcz-nixos-rebuild.drv
these paths will be fetched (1.39 MiB download, 7.12 MiB unpacked):
  /nix/store/lyhsl93f2w3vh7rf6plfvc6agvwzfdn5-nix-2.1.3
  /nix/store/xv9i3861lzjrbmg035465y6y86i33sql-nix-2.1.3-man
copying path '/nix/store/xv9i3861lzjrbmg035465y6y86i33sql-nix-2.1.3-man' from 'https://cache.nixos.org'...
copying path '/nix/store/lyhsl93f2w3vh7rf6plfvc6agvwzfdn5-nix-2.1.3' from 'https://cache.nixos.org'...
building '/nix/store/39vrdzsyw5wax59x6rmdk7hk1n0zlxcz-nixos-rebuild.drv'...
/nix/store/qfnl0zskb5sj1br41ndvav1pi0xs718p-nixos-rebuild
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=02b5a084-6302-4c4d-9268-663e27ee6ca4-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=3a66a3a0-d229-4947-82bb-00afad2c7d8b)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/hik4v1b8a8h3mga9g0ghx481kysjmqp5-nix-2.1.3-man' from 'https://cache.nixos.org'...
copying path '/nix/store/r7j7s6rakammqxrlpcrbyv5x8fdb8p17-brotli-1.0.6-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/ag088hzpkdlzd9lgxvi094bxicvcn32k-libatomic_ops-7.6.6' from 'https://cache.nixos.org'...
copying path '/nix/store/5fvlfkwzfl3a89apvvyaz7652clrcf76-aws-sdk-cpp-1.5.17' from 'https://cache.nixos.org'...
copying path '/nix/store/bwlaq8bvbpyqsfw9a5rgz7wr09rdpikm-curl-7.59.0' from 'https://cache.nixos.org'...
copying path '/nix/store/4ri26b7qgh53s65iqyj50sj4pvknl1qv-libsodium-1.0.16' from 'https://cache.nixos.org'...
copying path '/nix/store/846brfjv2rw3gk5yq744316qpkmk84jl-boehm-gc-7.6.8' from 'https://cache.nixos.org'...
copying path '/nix/store/crg1xmgsyqfcxk0wp5pycrzn32sbwwjh-nix-2.1.3' from 'https://cache.nixos.org'...
building '/nix/store/73qcwchc0aa4f9y35ywdrvy2l13lhhpx-nixos-rebuild.drv'...
/nix/store/q0asgrckwqgk4zl8y2z5v9dl2ws87isy-nixos-rebuild
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=02b5a084-6302-4c4d-9268-663e27ee6ca4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=2e232ce8-21c8-428a-88cb-9c7f7bba6cd4)

Attempted: nixos-rebuild

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/plwp97zfj51hgr3rv80jkqmwa8mdxv79-nixos-rebuild
```
</p></details>

zimbatm commented:
Seems like it's popular so I moved all the other tools.

There is a new `<nixpkgs/.codeName>` file so it's accessible from the pkgs side. I think we should introduce a `<nixpkgs/version.nix>` that holds the structured data but wanted to avoid breaking backward-compatibility.

`nixos-rebuild-vms` depends on `<nixpkgs/nixos/modules/installer/tools/nixos-build-vms/build-vms.nix>` which is not ideal in terms of data locality but I wanted to avoid breaking backward-compatibility here.

zimbatm commented:
@GrahamcOfBorg build nixos-build-vms nixos-enter nixos-generate-config nixos-install nixos-option nixos-rebuild nixos-version

Mic92 commented:
It seems that one could replace nixops with a shell script for the most part.

zimbatm commented:
I want to back-port this to 18.09. Did a bunch of tests and it seems to be working well.

Any objections?

zimbatm commented:
@GrahamcOfBorg build nixos-build-vms nixos-enter nixos-generate-config nixos-install nixos-option nixos-rebuild nixos-version

GrahamcOfBorg commented:
<!--REQUEST_ID=37c914d4-39b9-4a10-bf58-98c622dfcf70-->
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=f2bc3ac5-3060-45a8-99d3-22892de58d1e)

Attempted: nixos-build-vms, nixos-enter, nixos-install, nixos-option, nixos-rebuild, nixos-version

The following builds were skipped because they don't evaluate on x86_64-darwin: nixos-generate-config

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/m96ck5z05cwfrvnk12f7xj5rxbg6jadl-boehm-gc-7.6.8' from 'https://cache.nixos.org'...
copying path '/nix/store/wy3m25r5b75ihfpyvs48zjkvjdc3yf71-nix-2.1.3' from 'https://cache.nixos.org'...
building '/nix/store/dh65939vxnyq0zgy449w70i4k8lqyakb-nixos-install.drv'...
building '/nix/store/q924pq801rwwvdwnni4n2yiyf1q9mr16-nixos-rebuild.drv'...
/nix/store/q1qqyrvsliw15ddzvlyahm5w2y1qa8ix-nixos-build-vms
/nix/store/ps5wjqagp06s8bcn414lzmk9kdcr61ya-nixos-enter
/nix/store/gzqi4wlhyhlqjv7859fhrrapgpxfd0ws-nixos-install
/nix/store/gkg34xn55817kid22amlwsqmrxwn65d1-nixos-option
/nix/store/h4sdfj4cnvb42qslqpz7wkk4dr849q9x-nixos-rebuild
/nix/store/jcc18iqz600x82z5y03fk1qyqi0ckbfh-nixos-version
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=37c914d4-39b9-4a10-bf58-98c622dfcf70-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=960fedf8-f9c7-48c2-aa8a-6cfbddc9f433)

Attempted: nixos-build-vms, nixos-enter, nixos-generate-config, nixos-install, nixos-option, nixos-rebuild, nixos-version

<details><summary>Partial log (click to expand)</summary><p>

```
building '/nix/store/nlv0dfs6f4an6g6ljjafh5rrzg309dbd-nixos-generate-config.drv'...
building '/nix/store/jis1s6j0igj0i6sdjmqrf298c5aah8g8-nixos-install.drv'...
building '/nix/store/7ybs65wkbdd7qp4sx4vihfgzw20zbwkn-nixos-rebuild.drv'...
/nix/store/sj35sb30zwyqc94bsq6l2n0s42r8wzn4-nixos-build-vms
/nix/store/90lv72207g0jiwjvixxgcylrx5f9pzda-nixos-enter
/nix/store/ayrry8s855r04528swjf8f7c8zhlkfdd-nixos-generate-config
/nix/store/g4za42hdpkh8kq2w5v2chfz7g95h99wh-nixos-install
/nix/store/1niica3bk9l0ikn3qj1wmxp0c7ng0zvd-nixos-option
/nix/store/n9s3cawslna0nxv5i4my561r761px38v-nixos-rebuild
/nix/store/7ddj7pg17bjzkgrl7vkvsxz1kljwr4jd-nixos-version
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=37c914d4-39b9-4a10-bf58-98c622dfcf70-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.48122&attempt_id=55713417-9bdd-4a14-87d3-b7b7aab8b6da)

Attempted: nixos-build-vms, nixos-enter, nixos-generate-config, nixos-install, nixos-option, nixos-rebuild, nixos-version

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/lvl6zx25lhdmbc5sc9kcxq90z3apsm5h-nixos-rebuild' from 'https://cache.nixos.org'...
building '/nix/store/xxs5816zaa8jrl02jrsyqwhbj3qki7g3-nixos-build-vms.drv'...
building '/nix/store/zk27gl8hrryw2wwv22h20644nx3zz13l-nixos-version.drv'...
/nix/store/3pxa66im0zyrw64k2xxjmi6q6krbkqk4-nixos-build-vms
/nix/store/qd6a5wwl4370k60ci2wdqrsi0r0pajm5-nixos-enter
/nix/store/487ch5shz6bypi0saz6vxcwc5qgdfs1n-nixos-generate-config
/nix/store/x0qggva1a37iyi06hd3cp0zmvam3xrxb-nixos-install
/nix/store/0qnz2i2453znqjnd1hqf00nnzk5c2i3c-nixos-option
/nix/store/lvl6zx25lhdmbc5sc9kcxq90z3apsm5h-nixos-rebuild
/nix/store/fqb4m47abinn3qf88kl6hf3k4rssfgqx-nixos-version
```
</p></details>

zimbatm commented:
I'll just go ahead and merge this in master. If everything goes well, I will cherry-pick this in 18.09 next Monday.

edolstra commented on .codeName:
```diff
@@ -0,0 +1 @@
+Koi
```
Please don't use camelCase filenames. Also, does this really have to be in the root directory of the repository?

zimbatm replied:
how about something like this https://github.com/NixOS/nixpkgs/pull/48588 ?

edolstra commented:
I'm strongly against this. These are not packages - they should be in the NixOS source tree, not in `pkgs`.

zimbatm commented:
@edolstra is it a question of data locality, so that the script can evolve at the same pace as NixOS?

Personally I don't tend to make a bit distinction between both as they are in the same repo and have the same release schedule. NixOS is just another type of (albeit complex) derivation output.

Mic92 commented:
I would even say it is a complex derivation.

Ekleog commented:
@zimbatm At first I agreed with the idea that it'd be nice to have `nixos-rebuild` be a regular package, but it's true it makes sense only in the NixOS context, and thus has no meaning on all other systems.

So it'd require, to make sense, to either add a `platforms.nixos` platform, which sounds overkill… or to live in the `nixos/` tree.

Now, maybe a NixOS overlay to nixpkgs could be done, containing these? That'd solve both issues, with both the scripts as actual packages and packages in `nixos/`, explicitly being marked for NixOS.

zimbatm commented:
I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem.

After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.

Ekleog commented:
> I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem.
>
> After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.

Hmm… why not a nixos overlay put in the `nixos/` folder and enabled by
default on NixOS but pick-able on-demand on other systems? That would
sound like it'd also solve all the problems caused by `nixos-*` tools
not being packaged, to me

Mic92 commented:
> I talked with Eelco a bit and his main objection was to move the scripts outside of the nixos folder. It's a data locality problem. After NixCon I'll do another attempt where the packages are in pkgs/ but leaving the scripts in the nixos/ folder. And only port the ones that have an actual use-case.
> Hmm… why not a nixos overlay put in the `nixos/` folder and enabled by default on NixOS but pick-able on-demand on other systems? That would sound like it'd also solve all the problems caused by `nixos-*` tools not being packaged, to me

I think that can be achieved by using `(nixos{})` already.

Mic92 commented:
* easily boostrap nixos from a linux distribution + nix.

marler8997 commented:
I'm currently installing nix onto my second hard drive from Ubuntu.  I got the drive partitioned and the filesystems mounted, but then I was surprised to find that `nixos-generate-config` was not available as a package...?  Since this was removed...what's the recommended way forward for me here to complete the install?  Note that I don't have a CD drive.

Mic92 commented:
If you have USB port/stick, use dd the image to that. You can also use an kexec image that can drop you into a installer from your existing Linux: https://github.com/nix-community/nixos-generators

marler8997 commented:
Thanks for the tips, I didn't know about kexec I'll have to learn about that one.

So `nixos-generate-config` is still in nixpkgs, it's just not a top-level entry anymore.  Seems odd to a new Nix user like me.  I would think the reason not to put a tool in the top-level attribute set would be if it had a name that could clash...but `nixos-generate-config` is pretty explicit.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-is-nixos-rebuild-an-output-attribute-of-a-nixos-system/9904/2
